File: core/1221/User/Card.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Banking;

namespace Banking
{
    public class Card
    {
        private DateTime creationDate = DateTime.MinValue;
        private DateTime expirationDate = DateTime.MinValue;
        private String cardNumber = String.Empty; // 16 digits
        private SubBank subBank = null;

        internal DateTime CreationDate
        {
            get { return this.creationDate; }
            set { this.creationDate = value; }
        }
        internal DateTime ExpirationDate
        {
            get { return this.expirationDate; }
            set { this.expirationDate = value; }
        }
        internal String CardNumber
        {
            get { return this.cardNumber; }
            set { this.cardNumber = value; }
        }
        internal SubBank SubBank
        {
            get { return this.subBank; }
            set { this.subBank = value; }
        }
        internal Card()
        {
        }
        internal Card(DateTime creationDate)
        {
            this.creationDate = creationDate;
            this.expirationDate = creationDate.AddYears(5);
        }
        internal Card(DateTime creationDate, SubBank subBank) : this(creationDate)
        {
            this.subBank = subBank;
        }
        ~Card()
        {
        }
        private Int32 RandomNumberGenerator(Int32 length)
        {
            Random random = new Random();
            Int32 begin = (Int32)Math.Pow(10, length - 1);
            Int32 end = (Int32)Math.Pow(10, length);
            return random.Next(begin, end);

        }
        internal void Implement(String userID)
        {
            if (this.subBank != null)
            {
                this.cardNumber = "4" + userID.Substring(0, userID.Length - 3) + RandomNumberGenerator(8).ToString() + SubBank.GetID();
                this.creationDate = DateTime.Now;
                this.expirationDate = CreationDate.AddYears(5);

            }
        }
        internal void ExportInformation()
        {
            Console.WriteLine("Card information: ");
            Console.WriteLine("Sub bank: ", this.subBank.GetName());
            Console.WriteLine("Card number: ", this.cardNumber);
            Console.WriteLine("Creation date: ", this.creationDate.ToString("MM/yyyy"));
            Console.WriteLine("Expiration date: ", this.expirationDate.ToString("MM/yyyy"));
        }
    }
}

File: java/src/Card.java
import java.lang.Math;

import Interfaces.Exportable;
import SubBank.*;
import java.time.*;
import java.time.format.DateTimeFormatter;

public class Card implements Exportable {
    private String cardNumber = "";
    private LocalDateTime creationDate = LocalDateTime.MIN;
    private LocalDateTime expirationDate = LocalDateTime.MIN;
    private SubBank subBank = NullObject.getInstance();

    public String getCardNumber() {
        return this.cardNumber;
    }

    public void setCardNumber(String cardNumber) {
        this.cardNumber = cardNumber;
    }

    public LocalDateTime getCreationDate() {
        return this.creationDate;
    }

    public void setCreationDate(LocalDateTime creationDate) {
        this.creationDate = creationDate;
    }

    public LocalDateTime getExpirationDate() {
        return this.expirationDate;
    }

    public void setExpirationDate(LocalDateTime expirationDate) {
        this.expirationDate = expirationDate;
    }

    public SubBank getSubBank() {
        return this.subBank;
    }

    public void setSubBank(SubBank subBank) {
        this.subBank = subBank;
    }

    public Card() {

    }

    public Card(LocalDateTime creationDate) {
        this.creationDate = creationDate;
        this.expirationDate = creationDate.plusYears(5);
    }

    public Card(LocalDateTime creationDate, SubBank subBank) {
        this(creationDate);
        this.subBank = subBank;
    }

    private int randomNumberGenerator(int length) {
        int begin = (int) Math.pow(10, length - 1);
        int end = (int) Math.pow(10, length);
        return (int) Math.random() * (end - begin - 1) + begin;
    }

    public void create(String userID) {
        if (this.subBank.compareTo(NullObject.getInstance()) != 0) {
            this.cardNumber = "4" + userID.substring(0, userID.length() - 3) + randomNumberGenerator(8)
                    + subBank.getID();
            this.creationDate = LocalDateTime.now();
            this.expirationDate = creationDate.plusYears(5);
        }
    }

    public void exportInformation() {
        DateTimeFormatter uniqueFormatter = DateTimeFormatter.ofPattern("MM/yyyy");

        System.out.println("    Sub bank: " + this.subBank.getName());
        System.out.println("    Card number: " + this.cardNumber);
        System.out.println("    Creation date: " + uniqueFormatter.format(creationDate));
        System.out.println("    Expiration date: " + uniqueFormatter.format(creationDate));
    }
}
